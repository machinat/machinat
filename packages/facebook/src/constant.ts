export const FACEBOOK = 'facebook' as const;

export const MESSENGER_MESSAGING_TYPE_RESPONSE = 'RESPONSE';

export const PATH_MESSAGES = 'me/messages' as const;

export const PATH_PASS_THREAD_CONTROL = 'me/pass_thread_control' as const;
export const PATH_REQUEST_THREAD_CONTROL = 'me/request_thread_control' as const;
export const PATH_TAKE_THREAD_CONTROL = 'me/take_thread_control' as const;

export const PATH_MESSAGE_ATTACHMENTS = 'me/message_attachments' as const;
export const PATH_MESSENGER_PROFILE = 'me/messenger_profile' as const;
export const PATH_PERSONAS = 'me/personas' as const;

export const PATH_FEED = 'me/feed' as const;
export const PATH_PHOTOS = 'me/photos' as const;
export const PATH_VIDEOS = 'me/videos' as const;
