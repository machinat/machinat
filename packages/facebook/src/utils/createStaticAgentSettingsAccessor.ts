import { FacebookPageSettings } from '../types.js';
import { AgentSettingsAccessorI } from '../interface.js';

const createStaticAgentSettingsAccessor = (
  settings: FacebookPageSettings[],
): AgentSettingsAccessorI => ({
  getAgentSettings: async (page) =>
    settings.find(({ pageId }) => pageId === page.id) || null,
  getAgentSettingsBatch: async (pages) =>
    pages.map(
      (page) => settings.find(({ pageId }) => pageId === page.id) || null,
    ),
});

export default createStaticAgentSettingsAccessor;
