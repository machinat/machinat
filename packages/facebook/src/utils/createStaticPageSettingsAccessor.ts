import { FacebookPageSettings } from '../types';
import { PageSettingsAccessorI } from '../interface';

const createStaticPageSettingsAccessor = (
  settings: FacebookPageSettings[]
): PageSettingsAccessorI => ({
  getChannelSettings: async (page) =>
    settings.find(({ pageId }) => pageId === page.id) || null,
  getChannelSettingsBatch: async (pages) =>
    pages.map(
      (page) => settings.find(({ pageId }) => pageId === page.id) || null
    ),
  listAllChannelSettings: async () => settings,
});

export default createStaticPageSettingsAccessor;
