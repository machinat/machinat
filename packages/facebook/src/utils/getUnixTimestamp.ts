const getUnixTimestamp = (
  date: undefined | number | Date,
): undefined | number =>
  date instanceof Date ? Math.floor(date.getTime() / 1000) : date;

export default getUnixTimestamp;
