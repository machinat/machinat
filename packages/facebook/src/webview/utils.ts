/* eslint-disable import/prefer-default-export */
import type { ContextDetails } from '@sociably/auth';
import FacebookChat from '../Chat';
import FacebookUser from '../User';
import type { FacebookAuthContext, FacebookAuthData } from './types';

export const getAuthContextDetails = (
  data: FacebookAuthData
): ContextDetails<FacebookAuthContext> => {
  const { page: pageId, id } = data;
  return {
    user: new FacebookUser(pageId, id),
    thread: new FacebookChat(pageId, { id }),
    pageId,
  };
};
