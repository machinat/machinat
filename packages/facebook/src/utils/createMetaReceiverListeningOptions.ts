import { PopEventWrapper } from '@sociably/core';
import { ListeningPlatformOptions } from '@sociably/meta-api';
import FacebookBot from '../Bot.js';
import { FACEBOOK } from '../constant.js';
import eventFactory from '../event/factory.js';
import { FacebookEventContext } from '../types.js';

const createMetaReceiverListeningOptions = (
  bot: FacebookBot,
  popEventWrapper: PopEventWrapper<FacebookEventContext, null>
): ListeningPlatformOptions<FacebookEventContext> => ({
  platform: FACEBOOK,
  bot,
  objectType: 'page',
  makeEventsFromUpdate: (updateData) => {
    const { id: pageId, messaging, stanby } = updateData;
    const isStandby = stanby !== undefined;
    const rawEvents = isStandby ? stanby : messaging;

    return rawEvents.map((rawEvent) =>
      eventFactory(pageId, isStandby, rawEvent)
    );
  },
  popEvent: popEventWrapper(async () => null),
});

export default createMetaReceiverListeningOptions;
