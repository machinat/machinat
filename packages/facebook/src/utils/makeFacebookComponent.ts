import { makeNativeComponent } from '@sociably/core/renderer';
import { FACEBOOK } from '../constant.js';

export default makeNativeComponent(FACEBOOK);
