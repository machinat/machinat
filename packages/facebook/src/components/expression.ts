import { SociablyNode } from '@sociably/core';
import type { UnitSegment, OutputSegment } from '@sociably/core/renderer';
import { formatNode } from '@sociably/core/utils';
import makeFacebookComponent from '../utils/makeFacebookComponent';
import { PATH_MESSAGES } from '../constant';
import type {
  FacebookSegmentValue,
  FacebookComponent,
  MessageValue,
} from '../types';

/**
 * @category Props
 */
export type ExpressionProps = {
  /** Content nodes to be annotated. */
  children: SociablyNode;
  /**
   * The messaging type of the message being sent. For more information, see
   * [Sending Messages - Messaging Types](https://developers.facebook.com/docs/messenger-platform/send-messages/#messaging_types).
   * Default to 'RESPONSE'.
   */
  messagingType?: 'RESPONSE' | 'UPDATE' | 'MESSAGE_TAG';
  /**
   * The message tag to use when messagingType set to 'MESSAGE_TAG'. For more
   * information, see [Message Tags](https://developers.facebook.com/docs/messenger-platform/send-messages/message-tags).
   */
  tag?:
    | 'CONFIRMED_EVENT_UPDATE'
    | 'POST_PURCHASE_UPDATE'
    | 'ACCOUNT_UPDATE'
    | 'HUMAN_AGENT';
  /**
   * Push notification type.
   *  REGULAR: sound/vibration
   *  SILENT_PUSH: on-screen notification only
   *  NO_PUSH: no notification
   * Defaults to 'REGULAR'.
   */
  notificationType?: 'REGULAR' | 'SILENT_PUSH' | 'NO_PUSH';
  /**
   * Custom string that is delivered as a message echo. 1000 character limit.
   */
  metadata?: string;
  /** {@link QuickReply} elements to be sent with messages. */
  quickReplies?: SociablyNode;
  /** The persona id to use. */
  personaId?: string;
};

/**
 * Annotate all the children content with the message settings attributes and
 * append quick replies after the content.
 * @category Component
 * @props {@link ExpressionProps}
 * @guides Check official [doc](https://developers.facebook.com/docs/messenger-platform/send-messages)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/send-api).
 */
export const Expression: FacebookComponent<
  ExpressionProps,
  OutputSegment<FacebookSegmentValue>
> = makeFacebookComponent(async function Expression(
  {
    props: {
      children,
      messagingType,
      tag,
      notificationType,
      metadata,
      quickReplies,
      personaId,
    },
  },
  _path,
  render
) {
  const [childrenSegments, quickReplySegments] = await Promise.all([
    render<FacebookSegmentValue>(children, '.children'),
    render(quickReplies, '.quickReplies'),
  ]);

  if (childrenSegments === null) {
    return null;
  }

  const segments: OutputSegment<FacebookSegmentValue>[] = [];
  let lastMessageIdx = -1;

  for (const segment of childrenSegments) {
    if (segment.type === 'part') {
      throw new TypeError(
        `${formatNode(segment.node)} can not be placed in <Expression/>`
      );
    }

    // hoisting text to message object
    if (segment.type === 'text') {
      lastMessageIdx = segments.length;

      segments.push({
        type: 'unit',
        value: {
          type: 'message',
          apiPath: PATH_MESSAGES,
          params: {
            message: { text: segment.value },
            messaging_type: messagingType,
            tag,
            notification_type: notificationType,
            persona_id: personaId,
          },
        },
        node: segment.node,
        path: segment.path,
      });
    } else if (segment.type === 'unit' || segment.type === 'raw') {
      const { value } = segment;
      if (value.type === 'page' || value.type === 'comment') {
        throw new TypeError(
          `${formatNode(segment.node)} can not be placed in <Expression/>`
        );
      }

      const { apiPath, params, attachFile } = value;

      if (apiPath === PATH_MESSAGES && 'message' in value.params) {
        lastMessageIdx = segments.length;

        segments.push({
          ...segment,
          value: {
            type: 'message',
            apiPath,
            params: {
              ...value.params,
              messaging_type: messagingType,
              tag,
              notification_type: notificationType,
              persona_id: personaId,
            },
            attachFile,
          },
        });
      } else if (
        apiPath === PATH_MESSAGES &&
        'sender_action' in params &&
        (params.sender_action === 'typing_on' ||
          params.sender_action === 'typing_off')
      ) {
        segments.push({
          ...segment,
          value: {
            type: 'message',
            apiPath,
            params: { ...params, persona_id: personaId },
            attachFile: undefined,
          },
        });
      } else {
        segments.push(segment as UnitSegment<FacebookSegmentValue>);
      }
    } else if (segment.type !== 'break') {
      segments.push(segment);
    }
  }

  //  attach quick_replies and metadata to last message
  if (lastMessageIdx !== -1) {
    const lastMessageSeg = segments[
      lastMessageIdx
    ] as UnitSegment<MessageValue>;

    const { value } = lastMessageSeg;

    segments.splice(lastMessageIdx, 1, {
      ...lastMessageSeg,
      value: {
        ...value,
        params: {
          ...value.params,
          message: {
            ...value.params.message,
            metadata,
            quick_replies: quickReplySegments?.map((segment) => segment.value),
          },
        },
      },
    });
  }

  return segments;
});
