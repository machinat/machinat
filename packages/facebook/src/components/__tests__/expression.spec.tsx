import Sociably from '@sociably/core';
import { isNativeType } from '@sociably/core/utils';
import { PATH_MESSAGES } from '../../constant';
import { TextReply, EmailReply, PhoneReply } from '../quickReply';
import { Expression } from '../expression';
import { PassThreadControl } from '../handoverProtocol';
import { TypingOn, TypingOff, MarkSeen } from '../senderAction';
import { renderUnitElement } from './utils';

const quickReplies = (
  <>
    <TextReply title="foo" payload="bar" />
    <PhoneReply />
    <EmailReply />
  </>
);

const children = (
  <>
    <p>foo</p>
    bar
  </>
);

it('is valid component', () => {
  expect(typeof Expression).toBe('function');
  expect(isNativeType(<Expression>foo</Expression>)).toBe(true);
  expect(Expression.$$platform).toBe('facebook');
});

it.each([
  [
    'Expression with type',
    <Expression messagingType="UPDATE">{children}</Expression>,
  ],
  [
    'Expression with type and tag',
    <Expression messagingType="MESSAGE_TAG" tag="ACCOUNT_UPDATE">
      {children}
    </Expression>,
  ],
  [
    'Expression with notificationType',
    <Expression notificationType="NO_PUSH">{children}</Expression>,
  ],
  [
    'Expression with metadata',
    <Expression metadata="_SOME_METADATA_">{children}</Expression>,
  ],
  [
    'Expression with quickReplies',
    <Expression quickReplies={quickReplies}>{children}</Expression>,
  ],
  [
    'Expression with personaId',
    <Expression personaId="_PERSONA_ID_">{children}</Expression>,
  ],
  [
    'Expression with all props',
    <Expression
      messagingType="MESSAGE_TAG"
      tag="CONFIRMED_EVENT_UPDATE"
      notificationType="SILENT_PUSH"
      metadata="_MY_METADATA_"
      quickReplies={quickReplies}
      personaId="_PERSONA_ID_"
    >
      {children}
    </Expression>,
  ],
  [
    'Expression with break, pause and thunk in children',
    <Expression>
      <p>foo</p>
      bar
      <Sociably.Pause time={2000} />
      baz
      <Sociably.Thunk effect={async () => {}} />
    </Expression>,
  ],
])('%s match snapshot', async (_, element) => {
  await expect(renderUnitElement(element)).resolves.toMatchSnapshot();
});

it('returns null if empty children received', async () => {
  await expect(
    renderUnitElement(<Expression>{undefined}</Expression>)
  ).resolves.toBe(null);
  await expect(
    renderUnitElement(<Expression>{null}</Expression>)
  ).resolves.toBe(null);
});

it('hoist text value into message object', async () => {
  const segments = await renderUnitElement(
    <Expression>
      foo
      <p>bar</p>
      {'baz'}
    </Expression>
  );

  expect(segments).toMatchInlineSnapshot(`
    Array [
      Object {
        "node": "foo",
        "path": "$#Expression.children:0",
        "type": "unit",
        "value": Object {
          "apiPath": "me/messages",
          "params": Object {
            "message": Object {
              "text": "foo",
            },
            "messaging_type": undefined,
            "notification_type": undefined,
            "persona_id": undefined,
            "tag": undefined,
          },
        },
      },
      Object {
        "node": <p>
          bar
        </p>,
        "path": "$#Expression.children:1",
        "type": "unit",
        "value": Object {
          "apiPath": "me/messages",
          "params": Object {
            "message": Object {
              "text": "bar",
            },
            "messaging_type": undefined,
            "notification_type": undefined,
            "persona_id": undefined,
            "tag": undefined,
          },
        },
      },
      Object {
        "node": "baz",
        "path": "$#Expression.children:2",
        "type": "unit",
        "value": Object {
          "apiPath": "me/messages",
          "params": Object {
            "message": Object {
              "metadata": undefined,
              "quick_replies": undefined,
              "text": "baz",
            },
            "messaging_type": undefined,
            "notification_type": undefined,
            "persona_id": undefined,
            "tag": undefined,
          },
        },
      },
    ]
  `);
});

it('add attributes to message value', async () => {
  const segments = await renderUnitElement(
    <Expression
      messagingType="MESSAGE_TAG"
      tag="HUMAN_AGENT"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
    >
      foo
      <p>bar</p>
      {'baz'}
    </Expression>
  );

  segments!.forEach((segment) => {
    expect(segment.value).toEqual({
      apiPath: PATH_MESSAGES,
      params: {
        messaging_type: 'MESSAGE_TAG',
        tag: 'HUMAN_AGENT',
        notification_type: 'SILENT_PUSH',
        persona_id: '_PERSONA_ID_',
        message: {
          text: expect.any(String),
        },
      },
    });
  });
});

it('add persona_id to typeing_on/typeing_off sender action', async () => {
  const segments = await renderUnitElement(
    <Expression
      messagingType="MESSAGE_TAG"
      tag="ACCOUNT_UPDATE"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
    >
      foo
      <TypingOn />
      <TypingOff />
      <MarkSeen />
    </Expression>
  );

  expect(segments?.map(({ value }) => value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "message": Object {
            "metadata": undefined,
            "quick_replies": undefined,
            "text": "foo",
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "_PERSONA_ID_",
          "tag": "ACCOUNT_UPDATE",
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "persona_id": "_PERSONA_ID_",
          "sender_action": "typing_on",
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "persona_id": "_PERSONA_ID_",
          "sender_action": "typing_off",
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "sender_action": "mark_seen",
        },
        "type": "message",
      },
    ]
  `);
});

it('adds metadata to last message action', async () => {
  const segments = await renderUnitElement(
    <Expression metadata="_META_">
      foo
      <p>bar</p>
      <TypingOn />
    </Expression>
  );

  expect(segments?.map(({ value }) => value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "message": Object {
            "text": "foo",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "message": Object {
            "metadata": "_META_",
            "quick_replies": undefined,
            "text": "bar",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "persona_id": undefined,
          "sender_action": "typing_on",
        },
      },
    ]
  `);
});

it('adds quickReplies to last message action', async () => {
  const segments = await renderUnitElement(
    <Expression
      quickReplies={
        <>
          <TextReply title="foo" payload="bar" />
          <PhoneReply />
          <EmailReply />
        </>
      }
    >
      <p>foo</p>
      bar
      <TypingOn />
    </Expression>
  );

  expect(segments?.map(({ value }) => value)).toMatchInlineSnapshot(`
    Array [
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "message": Object {
            "text": "foo",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "message": Object {
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": "bar",
                "title": "foo",
              },
              Object {
                "content_type": "user_phone_number",
              },
              Object {
                "content_type": "user_email",
              },
            ],
            "text": "bar",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
      },
      Object {
        "apiPath": "me/messages",
        "params": Object {
          "persona_id": undefined,
          "sender_action": "typing_on",
        },
      },
    ]
  `);
});

it('do nothing to non-messgae value', async () => {
  const segments = await renderUnitElement(
    <Expression
      messagingType="MESSAGE_TAG"
      tag="CONFIRMED_EVENT_UPDATE"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
      quickReplies={<TextReply title="foo" payload="bar" />}
    >
      <PassThreadControl targetAppId={123} />
      foo
    </Expression>
  );

  expect(segments?.[0].value).toMatchInlineSnapshot(`
    Object {
      "apiPath": "me/pass_thread_control",
      "params": Object {
        "metadata": undefined,
        "target_app_id": 123,
      },
      "type": "message",
    }
  `);
  expect(segments?.[1].value).toMatchInlineSnapshot(`
    Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": Array [
            Object {
              "content_type": "text",
              "image_url": undefined,
              "payload": "bar",
              "title": "foo",
            },
          ],
          "text": "foo",
        },
        "messaging_type": "MESSAGE_TAG",
        "notification_type": "SILENT_PUSH",
        "persona_id": "_PERSONA_ID_",
        "tag": "CONFIRMED_EVENT_UPDATE",
      },
    }
  `);
});
