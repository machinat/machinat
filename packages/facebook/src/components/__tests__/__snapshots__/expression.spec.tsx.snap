// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expression with all props match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": "MESSAGE_TAG",
        "notification_type": "SILENT_PUSH",
        "persona_id": "_PERSONA_ID_",
        "tag": "CONFIRMED_EVENT_UPDATE",
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": "_MY_METADATA_",
          "quick_replies": Array [
            Object {
              "content_type": "text",
              "image_url": undefined,
              "payload": "bar",
              "title": "foo",
            },
            Object {
              "content_type": "user_phone_number",
            },
            Object {
              "content_type": "user_email",
            },
          ],
          "text": "bar",
        },
        "messaging_type": "MESSAGE_TAG",
        "notification_type": "SILENT_PUSH",
        "persona_id": "_PERSONA_ID_",
        "tag": "CONFIRMED_EVENT_UPDATE",
      },
      "type": "message",
    },
  },
]
`;

exports[`Expression with break, pause and thunk in children match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "bar",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": <Sociably.Pause
      time={2000}
    />,
    "path": "$#Expression.children:2",
    "type": "pause",
    "value": [Function],
  },
  Object {
    "node": "baz",
    "path": "$#Expression.children:3",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": undefined,
          "text": "baz",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": <Sociably.Thunk
      effect={[Function]}
    />,
    "path": "$#Expression.children:4",
    "type": "thunk",
    "value": [Function],
  },
]
`;

exports[`Expression with metadata match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": "_SOME_METADATA_",
          "quick_replies": undefined,
          "text": "bar",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
]
`;

exports[`Expression with notificationType match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": undefined,
        "notification_type": "NO_PUSH",
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": undefined,
          "text": "bar",
        },
        "messaging_type": undefined,
        "notification_type": "NO_PUSH",
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
]
`;

exports[`Expression with personaId match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": "_PERSONA_ID_",
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": undefined,
          "text": "bar",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": "_PERSONA_ID_",
        "tag": undefined,
      },
      "type": "message",
    },
  },
]
`;

exports[`Expression with quickReplies match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": Array [
            Object {
              "content_type": "text",
              "image_url": undefined,
              "payload": "bar",
              "title": "foo",
            },
            Object {
              "content_type": "user_phone_number",
            },
            Object {
              "content_type": "user_email",
            },
          ],
          "text": "bar",
        },
        "messaging_type": undefined,
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
]
`;

exports[`Expression with type and tag match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": "MESSAGE_TAG",
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": "ACCOUNT_UPDATE",
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": undefined,
          "text": "bar",
        },
        "messaging_type": "MESSAGE_TAG",
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": "ACCOUNT_UPDATE",
      },
      "type": "message",
    },
  },
]
`;

exports[`Expression with type match snapshot 1`] = `
Array [
  Object {
    "node": <p>
      foo
    </p>,
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "text": "foo",
        },
        "messaging_type": "UPDATE",
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children::1",
    "type": "unit",
    "value": Object {
      "apiPath": "me/messages",
      "params": Object {
        "message": Object {
          "metadata": undefined,
          "quick_replies": undefined,
          "text": "bar",
        },
        "messaging_type": "UPDATE",
        "notification_type": undefined,
        "persona_id": undefined,
        "tag": undefined,
      },
      "type": "message",
    },
  },
]
`;
